### PR TITLE
Testing Your Ruby Code - Enumerable

### PR DESCRIPTION
**Pull request by lirad**
-Installed Rspec
-Initialized Rspec and created a file for testing methods
-Added test cases to all methods
-Updated readme file

### CODE REVIEW

File: enumerable.rb [hunk 1]
```diff
@@ -25,7 +24,7 @@ def my_select
     my_each do |item|
       result.push(item) if yield(item)
     end
-    puts result
+    result
   end
 
   def my_all?(obj = nil)
```
Comment (lekegitrepo) [inline enumerable.rb:30]: - **_`#my_all?` when no block or argument is given returns true when none of the collection members are false or nil_**

- **_`#my_all?` when a class is passed as an argument returns false if any of the collection is not a member of such class_**

- **_`#my_all?` when a Regex is passed as an argument returns false if any of the collection does not match the Regex_**

File: enumerable.rb [hunk 2]
```diff
@@ -85,12 +86,13 @@ def my_count(obj = nil)
     counter
   end
 
-  def my_map(_array = nil)
+  def my_map(_obj = nil)
```
Comment (lekegitrepo) [inline enumerable.rb]: - **_`#my_map` executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum._**

File: enumerable.rb [hunk 3]
```diff
@@ -122,8 +124,12 @@ def my_inject(*obj)
   end
 end
```
Comment (lekegitrepo) [inline enumerable.rb:132]: - **_`#my_inject` raises a "LocalJumpError" when no block or argument is given_**

File: enumerable.rb [hunk 4]
```diff
@@ -122,8 +124,12 @@ def my_inject(*obj)
...
 
-def multiply_els(arr)
-  arr.my_inject(:*)
+class Multiply
+  def initialize; end
+
+  def multiply_els(arr)
+    arr.my_inject(:*)
+  end
 end
```
Comment (lekegitrepo) [inline enumerable.rb:140]: - **_`#multiply_els` accepts an array as an argument and multiplies all the elements of the array together using #my_inject_**

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
## Status: Request changes :recycle:
### You have done a good job :muscle:
#### 79 test cases, 68 pass :heavy_check_mark: and 11 fail :x:
#### Check my inline comments below and the comments I added here due to some comment glitch I was unable to add it directly to the affected section of the code.

- **_`#my_any?` when no block or argument is given returns true if at least one of the collection is not false or nil_**

- **_`#my_any?` when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric_**

- **_`#my_any?` when a Regex is passed as an argument returns true if any of the collection matches the Regex_**

- **_`#my_none?` when no block or argument is given returns true only if none of the collection members is true_**

- **_`#my_none?` when a Regex is passed as an argument returns false if any of the collection matches the Regex_**

**_Kindly fix these and submit them for review._**

#### _Reviewed by Temitayo Adeleke._

**Comment by lirad** [comment]
Hello @lekegitrepo, how are you? I just got your review on testing my ruby enumerable code but no remarks were made in the testing file. Just want to make sure I still have to work on the enumerates or if I should work on the test cases. Best,

File: spec/enumerables_spec.rb [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    it 'Checks for true if all elements matchs all the conditions' do
+      result = array_strings.my_all?(array_strings)
+      expect(result).to eq(true)
+    end
+
+    it 'Checks for false if all the conditions are not met' do
+      result = array_strings.my_all?(%w[John Mary Douglas])
+      expect(result).to eq(false)
+    end
+  end
```
Comment (maelfosso) [inline spec/enumerables_spec.rb:130]:
```rb


    it 'twist it a bit - 1' do
      true_array = [1, true, 'hi', []]
      expect(true_array.my_all?).eq(true_array.all?)

      false_array = [1, false, 'hi', []]
      expect(false_array.my_all?).eq(false_array.all?)
    end
```

File: spec/enumerables_spec.rb [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    it 'Checks if a element is selected from a hash' do
+      result = hash.my_select { |n| n['John'] }
+      compare = hash.select { |n| n['John'] }
+      expect(result).to eq(compare)
+    end
+  end
+
+  describe '#my_all?' do
+    it 'Checks for true if all elements matchs all the conditions' do
+      result = array_strings.my_all?(array_strings)
```
Comment (maelfosso) [inline spec/enumerables_spec.rb:106]: No added values. Change it please.

File: spec/enumerables_spec.rb [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    it 'Checks if returns true if any of the conditions are met' do
+      result = array_strings.my_any?('John')
+      compare = array_strings.any?('John')
+      expect(result).to eq(compare)
+    end
+    it 'Checks if returns false if none of the conditions are met' do
+      result = array_strings.my_any?('Pedro')
+      compare = array_strings.any?('Pedro')
+      expect(result).to eq(compare)
+    end
```
Comment (maelfosso) [inline spec/enumerables_spec.rb:161]:
```rb
it 'twist it a bit - 2' do
      true_array = [nil, false, true, []]
      expect(true_array.my_any?).eq(true_array.any?)

      false_array = [nil, false, nil, false]
      expect(false_array.my_any?).eq(false_array.any?)
    end
```

File: spec/enumerables_spec.rb [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+  describe '#my_none?' do
+    it 'Checks if returns true if none of the conditions are met' do
+      result = array_strings.my_none?('Pedro')
+      expect(result).to eq(true)
+    end
+
+    it 'Checks if returns false if any of the conditions are met' do
+      result = array_strings.my_none?('John')
+      expect(result).to eq(false)
+    end
```
Comment (maelfosso) [inline spec/enumerables_spec.rb:207]:
```rb
it 'twist it a bit - 3' do
      true_array = [nil, false, true, []]
      expect(true_array.my_none?).eq(true_array.none?)

      false_array = [nil, false, nil, false]
      expect(false_array.my_none?).eq(false_array.none?)
    end
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,
Thank you for applying the suggested changes.
However there are sill some issues to solve

1. I tweak a bit your tests and add some tests. They all failed. Read the comments below.

2. Add tests for `my_all?`, `my_any?`, `my_none?` for testing **String**, _**Integer**,_ **Regex**. Check also if the array is empty or contains the same values.

3. Please, expect `false` and `true`. Switch between the two. Test a method were the result will be **true** and somewhere else where the result will be **false**

Kindly make the suggested changes and submit a new code review request
Thanks

Mael FOSSO

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi Team    
#### Great Job so far, everything works fine, but there are a few more things required. 
### General
- [ ] Let's work on all changes required by @maelfosso 
- [ ] Lets test for when a class is passed as an argument.
- [ ] Please, expect false and true. Switch between the two. Test a method were the result will be true and somewhere else where the result will be false .**As Requested by the previous TSE**
### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

File: spec/enumerables_spec.rb [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+    it 'Checks if a element is selected from a hash' do
+      result = hash.my_select { |n| n['John'] }
+      compare = hash.select { |n| n['John'] }
+      expect(result).to eq(compare)
+    end
+  end
+
+  describe '#my_all?' do
+    it 'Checks for true if all elements in a integer array matchs all the conditions' do
+      result = array_integers.my_all?(array_integers.all?)
```
Comment (kobiyoyo) [inline spec/enumerables_spec.rb]: You are calling the main enumerable method inside yours as an argument isn't proper.
 `array_integers.my_all?(array_integers.all?)`
- [ ] You could check for when a block is given as an argument.



**Note:** Lets apply this to all `it` block in this scenario

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi Team    
#### Great Job so far, everything works fine, but there are a few more things required. 
### General
- [ ] Check comments below
### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Review by elmejdki — CHANGES_REQUESTED** [review]
# status: require changes :recycle: 

### Please work on the previously required changes before you submit another code review

**Review by elmejdki — APPROVED** [review]
# status: approved :heavy_check_mark: 
## Hi, it's me again, Great job thank you for making the required changes.

now you can merge, your project is approved :heavy_check_mark: 

## Happy Coding :tophat:

**Pull request merged**